### PR TITLE
Bonyuk Peter. Lab 4. Option 2.

### PR DESCRIPTION
**Pull request by PetrBonyuk**

### CODE REVIEW

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  StringRef getArgument() const final { return "bonyuk_fused_multiply_add"; }
+  StringRef getDescription() const final {
+    return "This Pass combines the operations of addition and multiplication "
+           "into one";
+  }
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    module.walk([&](Operation *operation) {
+      if (auto AddOperation = dyn_cast<LLVM::FAddOp>(operation)) {
```
Comment (aobolensk) [inline mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]: No need to walk through all operations, you can filter them by type in the walker itself

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: mlir/test/Transforms/lab4/bonyuk_peter/test.mlir [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+// void functiontwo(double a, double b){
+//     double c = 2 + a * b;
+// }
+
+// void functionthree(double a){
+//     double b = 2 + a * 8;
+// }
+
+module attributes {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<f80, dense<128> : vector<2xi32>>, #dlti.dl_entry<i64, dense<64> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr<272>, dense<64> : vector<4xi32>>, #dlti.dl_entry<i1, dense<8> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr, dense<64> : vector<4xi32>>, #dlti.dl_entry<i32, dense<32> : vector<2xi32>>, #dlti.dl_entry<i16, dense<16> : vector<2xi32>>, #dlti.dl_entry<i8, dense<8> : vector<2xi32>>, #dlti.dl_entry<!llvm.ptr<270>, dense<32> : vector<4xi32>>, #dlti.dl_entry<!llvm.ptr<271>, dense<32> : vector<4xi32>>, #dlti.dl_entry<f128, dense<128> : vector<2xi32>>, #dlti.dl_entry<f16, dense<16> : vector<2xi32>>, #dlti.dl_entry<f64, dense<64> : vector<2xi32>>, #dlti.dl_entry<"dlti.stack_alignment", 128 : i32>, #dlti.dl_entry<"dlti.endianness", "little">>} {
```
Comment (aobolensk) [inline mlir/test/Transforms/lab4/bonyuk_peter/test.mlir]:
```suggestion
module {
```

File: mlir/test/Transforms/lab4/bonyuk_peter/test.mlir [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    llvm.store %arg0, %1 {alignment = 8 : i64} : f64, !llvm.ptr
+    llvm.store %arg1, %2 {alignment = 8 : i64} : f64, !llvm.ptr
+    llvm.store %arg2, %3 {alignment = 8 : i64} : f64, !llvm.ptr
+    %4 = llvm.load %1 {alignment = 8 : i64} : !llvm.ptr -> f64
+    %5 = llvm.load %2 {alignment = 8 : i64} : !llvm.ptr -> f64
+    %6 = llvm.load %3 {alignment = 8 : i64} : !llvm.ptr -> f64 
+    %7 = llvm.intr.fma(%4, %5, %6) : (f64, f64, f64) -> f64 
+    // CHECK-NOT: %6 = llvm.fmul %4, %5 : f64
+    // CHECK-NOT: %7 = llvm.fadd %3, %6 : f64
+    llvm.return
```
Comment (aobolensk) [inline mlir/test/Transforms/lab4/bonyuk_peter/test.mlir:31]: Please add check for inserted fma instruction

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class BonyukFusedMultiplyAddPass
+    : public PassWrapper<BonyukFusedMultiplyAddPass, OperationPass<ModuleOp>> {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]: Please narrow down type of operation the pass is scheduled on.

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        if (auto MultiplyLeft = AddLeft.getDefiningOp<LLVM::FMulOp>()) {
+          HandleMultiplyOperation(AddOperation, MultiplyLeft, AddRight);
+        } else if (auto MultiplyRight =
+                       AddRight.getDefiningOp<LLVM::FMulOp>()) {
+          HandleMultiplyOperation(AddOperation, MultiplyRight, AddLeft);
+        }
+      }
+    });
+
+    module.walk([&](Operation *operation) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]:
```suggestion
    module.walk([](LLVM::FMulOp mulOp) {
```

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    Value FMAOperation = builder.create<LLVM::FMAOp>(
+        AddOperation.getLoc(), MultiplyOperation.getOperand(0),
+        MultiplyOperation.getOperand(1), Operand);
+    AddOperation.replaceAllUsesWith(FMAOperation);
+
+    if (MultiplyOperation.use_empty()) {
+      MultiplyOperation.erase();
+    }
+
+    if (FMAOperation.use_empty()) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]: Typo?

File: mlir/test/Transforms/lab4/bonyuk_peter/test.mlir [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    %1 = llvm.alloca %0 x f64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
+    %2 = llvm.alloca %0 x f64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
+    %3 = llvm.alloca %0 x f64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
+    llvm.store %arg0, %1 {alignment = 8 : i64} : f64, !llvm.ptr
+    llvm.store %arg1, %2 {alignment = 8 : i64} : f64, !llvm.ptr
+    llvm.store %arg2, %3 {alignment = 8 : i64} : f64, !llvm.ptr
+    %4 = llvm.load %1 {alignment = 8 : i64} : !llvm.ptr -> f64
+    %5 = llvm.load %2 {alignment = 8 : i64} : !llvm.ptr -> f64
+    %6 = llvm.load %3 {alignment = 8 : i64} : !llvm.ptr -> f64 
+    %7 = llvm.intr.fma(%4, %5, %6) : (f64, f64, f64) -> f64
```
Comment (m-ly4) [inline mlir/test/Transforms/lab4/bonyuk_peter/test.mlir]: Could you please add tests that has mul and add operations at input and fma on exit?

File: mlir/test/Transforms/lab4/bonyuk_peter/test.mlir [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    %2 = llvm.alloca %0 x f64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
+    %3 = llvm.alloca %0 x f64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
+    llvm.store %arg0, %1 {alignment = 8 : i64} : f64, !llvm.ptr
+    %4 = llvm.load %1 {alignment = 8 : i64} : !llvm.ptr -> f64
+    %5 = llvm.mlir.constant(2.0 : f64) : f64
+    %6 = llvm.fmul %4, %5 : f64
+    %7 = llvm.fadd %5, %6 : f64
+   // CHECK-NOT: %6 = llvm.intr.fma(%4, %5, %5) : (f64, f64, f64) -> f64
+   // CHECK-NOT: %7 = llvm.fadd %5, %6 : f64
+// CHECK: %6 = llvm.fmul %4, %5 : f64
```
Comment (m-ly4) [inline mlir/test/Transforms/lab4/bonyuk_peter/test.mlir]: Why your pass doesn't replace this case?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp (hunk 4 above)
Comment (m-ly4) [reply mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]: @PetrBonyuk Please apply this comment.

**Comment by PetrBonyuk** [comment]
This Pass is intended for operations representing floating point addition (LLVM::FAddOp) and floating point multiplication (LLVM::FMulOp). This pass combines these operations into one combined multiply-add operation.

File: mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp (hunk 4 above)
Comment (m-ly4) [reply mlir/lib/Transforms/lab4/bonyuk_peter/BonyukFusedMultiplyAddPass.cpp]: I meant to use FuncOp instead of ModuleOp.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**